### PR TITLE
Support repr(simd) on ADTs containing a single array field

### PR DESCRIPTION
**Pull request by gnzlbg**
This PR allows using `#[repr(simd)]` on ADTs containing a
single array field:

```rust
 #[repr(simd)] struct S0([f32; 4]);
 #[repr(simd)] struct S1<const N: usize>([f32; N]);
 #[repr(simd)] struct S2<T, const N: usize>([T; N]);
```

This should allow experimenting with portable packed SIMD
abstractions on nightly that make use of const generics.

It also has a fly-by cleanup that removes `simd_type` and `simd_size` methods.

r? @eddyb

### CODE REVIEW

File: src/librustc/ty/layout.rs
Comment (Centril) [inline src/librustc/ty/layout.rs]: Are these new post-monomorphization errors... if so, why that?

Comment (gnzlbg) [reply src/librustc/ty/layout.rs]: Not really, typeck errors before any of these can trigger, at least, as long as everything is correct there.

I added these to help during development. If typeck fails to error, an error here is better than an error down the line.

Comment (Centril) [reply src/librustc/ty/layout.rs]: Alright; this sounds good. Could you add comments to the places where you added these `.fatal` errors noting that these should be caught in typeck?

(Also, why isn't `bug!(...)` being used for ICEs here?)

Comment (gnzlbg) [reply src/librustc/ty/layout.rs]: > (Also, why isn't bug!(...) being used for ICEs here?)

No idea, the code was using fatal error for ICES already so I just also used that - the existing fatal error even has a test..

> Could you add comments to the places where you added these .fatal errors noting that these should be caught in typeck?

Sure.

Comment (gnzlbg) [reply src/librustc/ty/layout.rs]: So i've done that. 

The already existing error that fails if the element type isn't a machine type is currently not caught in typeck (this was already the case), and can fail if a generic SIMD vector is instantiated with an inappropriate element type. The way this is handled is that `libcore` should have a trait that bounds the accepted types so that this cannot happen.

There is a new monomorphization type error here now that we support generic lengths, and that's if the length passed to the vector is zero. We can only catch this once we know the actual value of the vector length, which right now is only at monomorphization time. 

Either libcore would bound the length of the vector, once we have bounds for const generics, or we should start accepting zero-element vectors (I don't see a reason to forbid this), but that is probably better done in a different PR.

Comment (gnzlbg) [reply src/librustc/ty/layout.rs]: > Either libcore would bound the length of the vector, once we have bounds for const generics, or we should start accepting zero-element vectors (I don't see a reason to forbid this), but that is probably better done in a different PR.

It's also unclear at this time whether we want to support non-powers-of-two; right now we support them, this PR does not change that.

Comment (eddyb) [reply src/librustc/ty/layout.rs]: > No idea, the code was using fatal error for ICES already so I just also used that - the existing fatal error even has a test..

`fatal` errors *are just errors* (that also abort the compilation early).
Only `(span_)bug` and panics are ICEs.

**Comment by eddyb** [comment]
cc @nagisa @rkruppe @alexcrichton

File: src/librustc_codegen_llvm/intrinsic.rs
Comment (eddyb) [inline src/librustc_codegen_llvm/intrinsic.rs]: `x` -> `ty` (or `simd_ty`)

File: src/librustc_codegen_llvm/intrinsic.rs
Comment (eddyb) [inline src/librustc_codegen_llvm/intrinsic.rs]: You could also do this with `layout.field(bx, 0).ty`, which seems cleaner.

File: src/librustc_codegen_llvm/intrinsic.rs
Comment (eddyb) [inline src/librustc_codegen_llvm/intrinsic.rs]: IMO it'd be better to leave this to `Abi::Vector`.

File: src/librustc_codegen_llvm/intrinsic.rs
Comment (eddyb) [inline src/librustc_codegen_llvm/intrinsic.rs]: Maybe rename this to `simd_elem_ty_and_count`?

File: src/librustc_codegen_llvm/intrinsic.rs
Comment (eddyb) [inline src/librustc_codegen_llvm/intrinsic.rs]: Might be a good idea to use the element `Scalar` from the `Abi::Vector` instead of the real type, but it would probably cause too much noise in this PR (if it can even work).

File: src/librustc_codegen_llvm/intrinsic.rs
Comment (eddyb) [inline src/librustc_codegen_llvm/intrinsic.rs]: Is this change even needed anymore?

File: src/librustc_typeck/check/mod.rs
Comment (eddyb) [inline src/librustc_typeck/check/mod.rs]: Now that I think about it, "machine type" is *such a weird name*, especially for something user-facing.

I think it comes from a time where `int` was a big int and `iN` was the "machine type" equivalent (but I'm not sure this ever happened). I'd rather it say "primitive scalar (integer/float/pointer)" (do we allow `bool` and `char`?).

cc @nikomatsakis

File: src/librustc_typeck/check/mod.rs
Comment (eddyb) [inline src/librustc_typeck/check/mod.rs]: Was this added later? It effectively invalidates the checks below...
Even more reason we should have a single `#[lang = "simd"]` instead of many `#[repr(simd)]`.

File: src/librustc_codegen_llvm/intrinsic.rs
Comment (gnzlbg) [reply src/librustc_codegen_llvm/intrinsic.rs]: Does Scalar give us the element type ? We need to at least distinguish between these types: `f{32,64}`, `{i,u}{8,16,32,64,128, size}`, `for<T> *mut T`, `for<T> *const T`.

File: src/librustc_codegen_llvm/intrinsic.rs
Comment (gnzlbg) [reply src/librustc_codegen_llvm/intrinsic.rs]: No, I don't think so. This was some downfall of the previous code that was calling the now removed `simd_size` and was passing `param_env` to it everywhere.

File: src/librustc_typeck/check/mod.rs
Comment (gnzlbg) [reply src/librustc_typeck/check/mod.rs]: I think this was added at some point last year to support generic vectors.

File: src/librustc_typeck/check/mod.rs
Comment (gnzlbg) [reply src/librustc_typeck/check/mod.rs]: > do we allow bool and char?

No, we do not. 

>  I'd rather it say "primitive scalar (integer/float/pointer)" 

I'd rather have a `is_simd_element_ty` check that just checks what we want to support for SIMD vectors.

File: src/librustc_codegen_llvm/intrinsic.rs
Comment (gnzlbg) [reply src/librustc_codegen_llvm/intrinsic.rs]: Where does `layout` come from ?

File: src/librustc_typeck/check/mod.rs
Comment (gnzlbg) [reply src/librustc_typeck/check/mod.rs]: For the time being at least we probably want to type check SIMD vectors early, e.g., `#[repr(simd)] struct Foo(Vec<i32>, Vec<i32>);` and `#[repr(simd)] struct Bar<const N: usize>([Vec<i32>;  N]);` should fail to type check. 

If we move to a single:

```rust
#[lang = "simd"]
struct Simd<T, const N: usize>([T; N]) where [T; N]: SomeTrait;
```

in the future then we might just accept anything and leave rejecting invalid inputs to `SomeTrait`.

File: src/librustc_codegen_llvm/intrinsic.rs
Comment (eddyb) [reply src/librustc_codegen_llvm/intrinsic.rs]: Really depends what it's used for. If it's for telling apart integer/float/pointer, then yeah, that's what `Scalar` does.

File: src/librustc_typeck/check/mod.rs
Comment (eddyb) [reply src/librustc_typeck/check/mod.rs]: I'm talking about the error message. For the internal method call, something like `is_valid_(for-)simd_element(_ty)` is fine (name's a bikeshed anyway).

File: src/librustc_codegen_llvm/intrinsic.rs
Comment (gnzlbg) [reply src/librustc_codegen_llvm/intrinsic.rs]: I don't see how I can use scalar to, e.g. tell apart a `*mut *mut usize` from a `*const i32`. 

Maybe you could show an example ?

Comment (eddyb) [reply src/librustc_codegen_llvm/intrinsic.rs]: No, you can't, but I don't know what the type is used for here.

File: src/librustc_codegen_llvm/intrinsic.rs
Comment (eddyb) [reply src/librustc_codegen_llvm/intrinsic.rs]: It's what you get from `layout_of(x)` below.

File: src/librustc_codegen_llvm/intrinsic.rs
Comment (gnzlbg) [reply src/librustc_codegen_llvm/intrinsic.rs]: > No, you can't, but I don't know what the type is used for here.

This function generates LLVM-IR for all the generic simd intrinsics. The LLVM-IR for packed SIMD vectors includes the element type, e.g., `<4 x float**>`, so we need to be able to tell whether we have a `float*`, `float**`, `float***`, etc.

Comment (gnzlbg) [reply src/librustc_codegen_llvm/intrinsic.rs]: We also use the `*const` vs `*mut` distinction, e.g., when generating monomorphization errors, e.g., we don't want to allow writes via `*const`.

Comment (eddyb) [reply src/librustc_codegen_llvm/intrinsic.rs]: Look in `type_of.rs` for how that's done, I guess? I don't remember exactly, but we could offer something similar to `scalar_pair_element_llvm_type`, except for vectors.

I guess this is the kind of place where a signature like `ptrs: Simd<*mut T, N>, vals: Simd<T, N>` would be *really nice*, since you can rely on the typesystem for correctness.

Oh, bonus: `shuffle(a: Simd<T, A>, b: Simd<T, B>, indices: [usize; N]) -> Simd<T, N>` :D.

File: src/librustc/ty/layout.rs
Comment (gnzlbg) [reply src/librustc/ty/layout.rs]: @eddyb Is `fatal` the preferred way to omit monomorphization time errors?

Comment (eddyb) [reply src/librustc/ty/layout.rs]: Yes, although ideally there shouldn't be any...

Comment (gnzlbg) [reply src/librustc/ty/layout.rs]: @centril so is this issue resolved?

Comment (Centril) [reply src/librustc/ty/layout.rs]: I leave that to @eddyb -- the only thing I would note is that we don't want to add new monomorphization errors in stable but such a consideration seems far off since this is just for experimentation.

Comment (gnzlbg) [reply src/librustc/ty/layout.rs]: There is a huge difference between monomorphization time errors that affect Rust users, and adding code to catch logic errors in rustc early. 

I regularly get LLVM errors when using Rust. Their output is completely useless, and the first step into debugging those is requiring users to re-compile a Rust toolchain with LLVM assertions enabled, and debugging back to Rust from there. 

I'd much rather have a monomorphization time error just saying "This generic intrinsics does not support that type", and wish others would as well.

Comment (eddyb) [reply src/librustc/ty/layout.rs]: I think I know what needs to be cleared up: the `fatal` was moved from one place to another, it's not new (AFAIK).

Comment (Centril) [reply src/librustc/ty/layout.rs]: @gnzlbg I think you misunderstood me; I was making a general remark about user exposed monmorphization time errors, not internal ones for intrinsics... so we should be good.

File: src/librustc_codegen_llvm/intrinsic.rs
Comment (gnzlbg) [reply src/librustc_codegen_llvm/intrinsic.rs]: We discussed this on Discord where I think the conclusion was that we do need to obtain the exact Rust field type to deal with some of the intrinsics, right?

File: src/librustc_typeck/check/mod.rs
Comment (gnzlbg) [reply src/librustc_typeck/check/mod.rs]: I've improved the error messages. The internal method is still called `.is_machine`.

**Comment by gnzlbg** [comment]
So apparently `align 4` is not generated on Linux but is generated on osx, I need to skip that. The important part of the test is to verify that the memcpy actually copies the whole vector.

File: src/librustc_codegen_llvm/intrinsic.rs
Comment (eddyb) [reply src/librustc_codegen_llvm/intrinsic.rs]: Yeah I think we just need to postpone this on having a `Simd<T, N>` type.

Comment (gnzlbg) [reply src/librustc_codegen_llvm/intrinsic.rs]: So I've opened an issue to track doing this: https://github.com/rust-lang/rust/issues/63633

File: src/librustc/ty/layout.rs
Comment (Dante-Broggi) [inline src/librustc/ty/layout.rs]: Typo: “isn;t” => “isn’t”

**Comment by JohnCSimon** [comment]
ping from triage 
@gnzlbg can you please resolve the typo? 
thank you

**Comment by eddyb** [comment]
Do we need a lang item sign-off on this? I'm not sure of the procedures here.

**Comment by JohnCSimon** [comment]
Ping from triage - 
@gnzlbg @eddy is this ready to go into review? 
thanks

**Comment by gnzlbg** [comment]
I think I addressed all issues.

**Review by eddyb — APPROVED** [review]
r=me modulo https://github.com/rust-lang/rust/pull/63531#issuecomment-524587895

**Comment by gnzlbg** [comment]
@Centril ^^^

**Comment by Mark-Simulacrum** [comment]
Nominated for lang team and marked as waiting on team. It may be better to FCP this PR, though.

**Comment by Centril** [comment]
👍 I'm happy to sign off on this on a "let's experiment basis" (per FCP) without any other commitments. :)

**Comment by scottmcm** [comment]
@gnzlbg Now that the FCP's done, can you rebase please?

**Comment by JohnCSimon** [comment]
Ping from triage.
@gnzlbg - please address the merge conflicts 
cc: @scottmcm @eddyb 
Thank you!

**Comment by JohnCSimon** [comment]
Pinging again from triage.
@gnzlbg - can you please rebase and address the merge conflicts?
cc: @scottmcm @eddyb
Thank you!

**Comment by JohnCSimon** [comment]
Pinging again from triage....
@gnzlbg - can you please rebase and address the merge conflicts?
cc: @scottmcm @eddyb @Centril 
Thank you!

**Comment by Mark-Simulacrum** [comment]
While unfortunate, unless someone picks this up and finishes the rebase, we'll likely be closing it when the next round of triage comes around. @gnzlbg -- can you provide at least some feedback or sense of timeline on when we can expect progress here? Maybe we can ask for help from someone who's familiar with this patch who can help rebase it?

**Comment by dlight** [comment]
@Mark-Simulacrum :( isn't closing it a waste of review and implementation effort? This is an useful feature, all this discussion for nothing..

I petition to keep it open until it's rebased and merged

**Comment by Centril** [comment]
@dlight The PR can always be reopened when someone has the time to finish the work. For example, you could pick it up and finish it yourself.

**Comment by gnzlbg** [comment]
I should have time to finish it this afternoon, I started rebasing a couple of time but got too many conflicts.

**Comment by JohnCSimon** [comment]
Ping from triage
@gnzlbg Just letting you know this has sat idle for awhile. 
Thank you!

**Comment by JohnCSimon** [comment]
Ping from triage
@gnzlbg Just letting you know this has sat idle for awhile 
can you please address merge conflicts
cc: @Centril @eddyb 
Thank you!

**Comment by JohnCSimon** [comment]
Pinging again from triage
@gnzlbg This PR has sat idle for awhile. Please address merge conflicts
cc: @Centril @eddyb
Thank you!

**Comment by JohnCSimon** [comment]
Hello from triage
Unfortunately this one has sat idle for far too long and I'm marking this as inactive. 
Feel free to reopen once you have the time.
@gnzlbg @dlight @eddyb @Mark-Simulacrum

**Pull request closed without merging**
